feat: Scale label menu, connect Summary-Emotions-Others inputs to BE

teddy-f-47 opened the pull request:
I am sorry for the big size of the PR. 😅 Below I try to divide the list of changes into four topics.

# feature: scale labels import menu
What's changed:
 - frontend/.eslintrc.js : Increased line length limit from 100 to 150
 - frontend/plugins/services.ts : Added `scaleType` service which uses existing `LabelApplicationService` and `APILabelRepository`
 - frontend/pages/projects/_id/labels/index.vue : Added `scale` menu for affective annotation project
 - frontend/pages/projects/_id/labels/import.vue : Added scaleType service for importing scale labels
 - frontend/pages/projects/_id/labels/add.vue : Added scaleType service for adding scale labels manually
 - frontend/pages/projects/_id/labels/_label_id/edit.vue : Added scaleType service for editing scale labels manually
What's new:
 - No new file, but there is a new service called scaleType (accessible via i.e. `this.$services.scaleType`) for importing scale labels and also for retrieving scale labels later. The method `this.$services.scaleType.list()` will return a list of LabelDTO objects. For this service, the most important fields in LabelDTO are `id` (which is later needed as the `label_id` when submitting scale annotation) and `text` (which corresponds to the imported scale label).
Screenshot:
<img width="960" alt="ss-scale" src="https://user-images.githubusercontent.com/64476430/195352867-6523285d-4d36-42b4-9f73-9414b88e7652.PNG">


# feature: affectiveTextlabel service for saving textfield-based data into table `labels_textlabel` in db
What's new:
 - frontend/domain/models/tasks/affectiveTextlabel.ts : Added the model for data that will be stored in table `labels_textlabel` in db. It is actually quite similar to the seq2seq model, but with the additional `question` field.
 - frontend/repositories/tasks/affectiveAnnotation/apiAffectiveTextlabel.ts : Added the api endpoint for storing data in table `labels_textlabel` in db. It is actually the same api endpoint as the one in seq2seq task, but I decided to separate it because it is a different task.
 - frontend/services/application/tasks/affectiveAnnotation/affectiveTextlabelData.ts : Added AffectiveTextlabelDTO class.
 - frontend/services/application/tasks/affectiveAnnotation/affectiveTextlabelApplicationService.ts : Added AffectiveTextlabelApplicationService for retrieving and sending textfield-based data from/to BE.
What's changed:
 - frontend/plugins/services.ts : Registered a new service called affectiveTextlabel. We can use `this.$services.affectiveTextlabel.list(projectId, docId)` to retrieve data from `labels_textlabel` table in db for a given article text. We can use `this.$services.affectiveTextlabel.create(projectId, docId, answerString, questionString)` to save data into `labels_textlabel` table in db.
Screenshots:
<img width="960" alt="ss-summary1" src="https://user-images.githubusercontent.com/64476430/195353047-9bcfabc3-da04-4cbd-bcc2-f2a2dab0f7aa.PNG">
<img width="672" alt="ss-summary2" src="https://user-images.githubusercontent.com/64476430/195353050-7dd1d3b4-9a5d-440e-8b68-2e33412f44c6.PNG">


# feature: affectiveScale service for saving slider-based data into table `labels_scale` in db
What's new:
 - frontend/domain/models/tasks/affectiveScale.ts : Added the model for scale data that will be stored in table `labels_scale` in db.
 - frontend/repositories/tasks/affectiveAnnotation/apiAffectiveScale.ts : Added the api endpoint for storing scale data in table `labels_scale` in db.
 - frontend/services/application/tasks/affectiveAnnotation/affectiveScaleData.ts : Added AffectiveScaleDTO class.
 - frontend/services/application/tasks/affectiveAnnotation/affectiveScaleApplicationService.ts : Added AffectiveScaleApplicationService for retrieving and sending slider-based data from/to BE.
What's changed:
 - frontend/plugins/services.ts : Registered a new service called affectiveScale. We can use `this.$services.affectiveScale.list(projectId, docId)` to retrieve scale data from `labels_scale` table in db for a given article text. We can use `this.$services.affectiveScale.create(projectId, docId, labelId, scaleValue)` to save scale data into `labels_scale` table in db.
Screenshots:
<img width="960" alt="ss-slider1" src="https://user-images.githubusercontent.com/64476430/195353099-98856bf2-ff70-438e-bae4-c1a9031b1cad.PNG">
<img width="672" alt="ss-slider2" src="https://user-images.githubusercontent.com/64476430/195353105-43ceb4be-ed2c-40a1-94dc-b15f039777b7.PNG">


# Adjustment on affective annotation page
What's changed:
 - frontend/components/tasks/affectiveAnnotation/others/OthersInput.vue : Minor adjustment in handling changes in the textfieldwithseq2seq component.
 - frontend/components/tasks/affectiveAnnotation/inputs/Slider.vue : Minor adjustment with the slider thumb size setting.
 - frontend/pages/projects/_id/affective-annotation/index.vue : Refactor all scale data to be put inside `affectiveScalesValues`, in `async created()` loaded the list of scale labels in the project with id-to-text mapping and text-to-id mapping, in `async list(docId)` loaded textlabel and scale data. Finally, the Summary, Emotions, and Others inputs have been connected to BE.


Notable:
 - Before merging this PR, the PR #34 should be merged first in order to ensure data integrity in db.

teddy-f-47 commented:
This is the scale labels JSON file that I used to import:
```
[
	{
        "text": "Pozytywne",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
    {
        "text": "Negatywne",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
    {
        "text": "Radość",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Zachwyt",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
    {
        "text": "Inspiruje",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Spokój",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
    {
        "text": "Zaskoczenie",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Współczucie",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
    {
        "text": "Strach",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Smutek",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
    {
        "text": "Wstręt",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Złość",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Ironiczny",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Żenujący",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Wulgarny",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Polityczny",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Interesujący",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Zrozumiały",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Zgadzam się z tekstem",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Wierzę w tę informację",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Czuję sympatię do autora",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    },
	{
        "text": "Potrzebuję więcej informacji, aby ocenić ten tekst",
        "suffix_key": "",
        "background_color": "#FF0000",
        "text_color": "#ffffff"
    }
]
```

artikandri commented on frontend/pages/projects/_id/affective-annotation/index.vue:
```diff
@@ -371,6 +353,20 @@ export default {
+      this.scaleTypesIdsTexts = scaleTypesIdsTexts
+      this.scaleTypesTextsIds = scaleTypesTextsIds
+    } else {
+      alert("No scale labels found. Please import scale labels first!")
```
I wonder, rather than giving an alert, how about just making it so that the code will not work when there is no scales present or give a warning in form of a text? Because the alert has never been used in the rest of the system, as long as I can remember  XD Just an idea

teddy-f-47 replied:
1. I agree, ideally the JSON file containing the scale labels should be available somewhere in `frontend` directory so that it can be imported to the db automatically and also the same file can be used in `index.vue`. In that case, the mapping can refer to the content of the file and everything will be more-or-less fine XD Do you think I should just add the JSON file in this PR?
2. Yeah XD I just use alert as a temporary solution for now..... And it is not really good also because the Summary mode doesn't need scale labels, but the alert will appear too. XD Maybe we can add an additional variable like `isScaleImported`, and if the value is false then the annotation box won't be shown, do you like that approach?

I think I will be able to make further adjustments on Monday 😅

artikandri replied:
1. I think that will be great, yes. 
2. Yes, that will be nice. XD 

Will be waiting for the adjustments.

artikandri reviewed (commented):
Hello, thank you for the good work! I have tested it, and it works well on my side. 

As for the hardcoded Polish language, in my case, I will handle it by importing the Polish annotation file to the `index.vue` and then accessing it by the index, which is possible for me because I wrote my code to be array-based and I wrote all the keys in some patterns, but I don't know if the approach will work here. 

As for now, since I got nothing to comment other than the `alert`, everything looks good for me. XD  Was about to comment the `console.log` yesterday but you already removed it.

teddy-f-47 commented:
Hi all, I have updated the PR. I have addressed the two points in Argi's comment, so:
(1) I have added the scale labels JSON files for the two modes that I'm handling, I put them in the `static/formats/affective_annotation/` directory. Notably I added the field `varname` in each element, which won't be imported into db on BE side, but I added it to make the mapping easier. I have some ideas about refactoring the input component to make the code a bit more cleaner without relying on such mapping, but I think I prefer to stick with the current working code for now, and do the refactoring later.
(1.5) I removed `json` from the list of filetypes to be loaded with the raw-loader, at first the JSON files were causing me errors and I tried to debug for hours until finally I found this: [issue](https://github.com/webpack/webpack/issues/6586), and the recommended solution there.
(2) I have removed the ugly js alert XD and now we just don't show the annotation box if the scales labels haven't been imported and the mode is not summary mode. However, after Offensive and Humor inputs are merged, I will rebase and refactor the code inside index.vue to conform with Argi's code (the v-if will probably be moved somewhere else).

haanh764 reviewed (approved):
Thank you for your great work, I think everything looks fine now XD Approved!
Btw, I have only 1 minor comment, not a comment but a question XD that in the "Labels" section, we still have the Category and Span there which could be quite confused, because we don't have those functions in this project type. So should we remove it (?)

teddy-f-47 commented:
> Thank you for your great work, I think everything looks fine now XD Approved! Btw, I have only 1 minor comment, not a comment but a question XD that in the "Labels" section, we still have the Category and Span there which could be quite confused, because we don't have those functions in this project type. So should we remove it (?) And remember to solve the conflict as well XD

I am not sure if we should remove it or not, but most probably yes -- I think the affective annotation will not use category, span, or relation labels. If everyone agrees on this, then I will also remove Category and Span options in the Label menu XD Sure, I will rebase and solve the conflict XD

haanh764 commented:
> > Thank you for your great work, I think everything looks fine now XD Approved! Btw, I have only 1 minor comment, not a comment but a question XD that in the "Labels" section, we still have the Category and Span there which could be quite confused, because we don't have those functions in this project type. So should we remove it (?) And remember to solve the conflict as well XD
> 
> I am not sure if we should remove it or not, but most probably yes -- I think the affective annotation will not use category, span, or relation labels. If everyone agrees on this, then I will also remove Category and Span options in the Label menu XD Sure, I will rebase and solve the conflict XD

Yes, I think we can remove it later if everyone agree on this XD just note it haha

teddy-f-47 commented:
I have rebased and fixed the conflicts from recent merges. :)

artikandri reviewed (approved):
Hello! I have tested the branch and it looks fine, thank you for your work! I will wait for the refactor later :) 

And as for the discussion above, I agree; we can hide the tables if they are not used.

artikandri reviewed (approved):
Hello! I have tested the branch and it looks fine, thank you for your work! I will wait for the refactor later :) 

And as for the discussion above, I agree; we can hide the tables if they are not used.
